### PR TITLE
Digest Authentication

### PR DESCRIPTION
**Pull request by dzervas**
### 1. What does this change do, exactly?
It adds a `digestauth` directive that works exactly like `basicauth` and implements (guess what) digest authentication.

### 2. Please link to the relevant issues.
#712 

### 3. Which documentation changes (if any) need to be made because of this PR?
Add the `digestauth` directive and deprecate advise against basic authentication for security reasons.

### 4. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I am willing to help maintain this change if there are issues with it later

### CODE REVIEW

**Comment by dzervas** [comment]
I'm kinda in a dead end here. I need help to write the tests (how can I keep state of the nonces and the whole handshaking?).

**Comment by mholt** [comment]
Ooo, thanks for this! Been a busy week this week, I will take a look at this soon! Anyone else is welcome to contribute as well.

File: caddyhttp/digestauth/parse.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	"log"
+	"strings"
+	"testing"
+)
+
+// Tear apart the Authorization header value.
+// PAY ATTENTION: this is large and complicated relative to other ones I've seen
+// based on Split() using ',', ' ', and '=' in various orders. It is also probably
+// correct even if the realm contains a '=', ' ', or '"' character, or if the
+// sender uses HT, CR, or LF in their whitespace.
```
Comment (mholt) [inline caddyhttp/digestauth/parse.go]: Impressive. This probably merits some testing.

File: caddyhttp/digestauth/setup.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
+package digestauth
+
+import (
+	"strings"
+	//"syscall"
```
Comment (mholt) [inline caddyhttp/digestauth/setup.go]: Syscall?

File: caddyhttp/digestauth/setup_test.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+				Resources: []string{"/res2"}, Realm: "Restricted"},
+		}},
+		{`digestauth user`, true, "", []Rule{}},
+		{`digestauth`, true, "", []Rule{}},
+		{`digestauth /resource user pwd asdf`, true, "", []Rule{}},
+
+		//{`digestauth sha1 htpasswd=` + htfh.Name(), false, htpasswdPasswd, []Rule{
+		//{Users: NewHtdigestUserStore(htfh.Name(), nil),
+		//Resources: []string{"/res2"}, Realm: "Restricted"},
+		//}},
```
Comment (mholt) [inline caddyhttp/digestauth/setup_test.go]: Why is this commented out?

File: caddyhttp/digestauth/setup_test.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+				continue
+			}
+			pwd := test.password
+			if len(actual) > 1 {
+				pwd = fmt.Sprintf("%s%d", pwd, j+1)
+			}
+			//if !actualRule.Password(pwd) || actualRule.Password(test.password+"!") {
+			//t.Errorf("Test %d, rule %d: Expected password '%v', got '%v'",
+			//i, j, test.password, actualRule.Password(""))
+			//}
```
Comment (mholt) [inline caddyhttp/digestauth/setup_test.go]: (And this one)

File: caddyhttp/digestauth/digestauth_test.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+	//filename := filepath.Base(htfh.Name())
+	//if rule.Password, err = GetHtpasswdMatcher(filename, rule.Username, siteRoot); err != nil {
+	//t.Fatalf("GetHtpasswdMatcher(%q, %q): %v", htfh.Name(), rule.Username, err)
+	//}
+	//t.Logf("%d. username=%q", i, rule.Username)
+	//if !rule.Password(htpasswdPasswd) || rule.Password(htpasswdPasswd+"!") {
+	//t.Errorf("%d (%s) password does not match.", i, rule.Username)
+	//}
+	//}
+}
```
Comment (mholt) [inline caddyhttp/digestauth/digestauth_test.go]: Is this related to the other commented tests?

File: caddyhttp/digestauth/setup.go (hunk 2 above)
Comment (dzervas) [reply caddyhttp/digestauth/setup.go]: oops, left there by mistake

File: caddyhttp/digestauth/setup_test.go (hunk 3 above)
Comment (dzervas) [reply caddyhttp/digestauth/setup_test.go]: cause the tests are messed up. I tried to have "something" working as a start point. I left it for 2-3 weeks and had no idea what I wanted to do... :/

File: caddyhttp/digestauth/setup_test.go (hunk 4 above)
Comment (dzervas) [reply caddyhttp/digestauth/setup_test.go]: ^ same

File: caddyhttp/digestauth/digestauth_test.go (hunk 5 above)
Comment (dzervas) [reply caddyhttp/digestauth/digestauth_test.go]: ^ same

File: caddyhttp/digestauth/digestauth.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
+// Package digestauth implements HTTP Digest Authentication for Caddy.
+//
+// This is useful for simple protections on a website, like requiring
+// a password to access an admin interface. This package assumes a
+// fairly small threat model.
```
[the remark this answers is not in the document]
Comment (dzervas) [reply caddyhttp/digestauth/digestauth.go]: low risk

File: caddyhttp/digestauth/digest.go [hunk 7]
```diff
@@ -0,0 +1,297 @@
+/*
+Package digest provides HTTP Digest authentication.
+
+You may use Apache style htdigest files,
+a simple map of userXrealm->MD5(user:realm:password) or your own implementation of a user
+lookup function (if perhaps you use a database).
+
+The API is designed for HTTP request routers like Martini. It provides a handler, which
```
[the remark this answers is not in the document]
Comment (dzervas) [reply caddyhttp/digestauth/digest.go]: I got a lot of code from [here](https://github.com/jimstudt/http-authentication/tree/master/digest). I couldn't just `include` it, as it populates only the setup function for martini and this does not fit caddy. I'm not sure if this still works with martini.

File: caddyhttp/digestauth/digest.go [hunk 8]
```diff
@@ -0,0 +1,296 @@
+/*
+Package digest provides HTTP Digest authentication.
```
Comment (bradfitz) [inline caddyhttp/digestauth/digest.go]: Package digestauth, not digest

File: caddyhttp/digestauth/digest.go [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+You may use Apache style htdigest files,
+a simple map of userXrealm->MD5(user:realm:password) or your own implementation of a user
+lookup function (if perhaps you use a database).
+
+The API is designed for HTTP request routers like Martini. It provides a handler, which
+will do nothing to the HTTP response for valid authentication. For invalid authentication
+it will make an http.Error(), which will usually be a 401 (Unauthorized) to request
+authentication from a user.
+
+You might use digest like this if you were getting passwords from your own source:
```
Comment (bradfitz) [inline caddyhttp/digestauth/digest.go]: digestauth? or just "Example usage:"

File: caddyhttp/digestauth/digest.go [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+     }
+     ...
+     digester := auth.NewDigestHandler( "My Realm", nil, nil, myUserFile )
+     ...
+     m.Post("/my-sensitive-uri", digester.ServeHTTP, mySensitiveHandler)  // just protect this one, notice chained handlers.
+
+You will have noticed that both New*Handler calls included a pair of
+nils. If you don't like the way nonces are created, tracked, and expired,
+then you will want to replace one or both of these with your own
+implementations of Nonce and NonceStore.  Otherwise, ignore those
```
Comment (bradfitz) [inline caddyhttp/digestauth/digest.go]: More idiomatic is to not take optional parameters in constructors and make them exported fields instead that users can optionally set to be non-nil before use.

File: caddyhttp/digestauth/digest_test.go [hunk 11]
```diff
@@ -0,0 +1,172 @@
+// Package digest provides HTTP Digest authentication. You may use Apache style htdigest files,
```
Comment (bradfitz) [inline caddyhttp/digestauth/digest_test.go]: copy/paste mistake? package doc not needed in your test file.

File: caddyhttp/digestauth/simple-user-store.go [hunk 12]
```diff
@@ -0,0 +1,25 @@
+package digestauth
+
+import ()
```
Comment (bradfitz) [inline caddyhttp/digestauth/simple-user-store.go]: delete

**Comment by mholt** [comment]
Pinging @dzervas -- do you think you'll be able to help finish up this PR in the next few weeks?

**Comment by dzervas** [comment]
Unfortunately no. (I have to relocate, got a job, etc.)
I'm sorry!

**Comment by mholt** [comment]
@dzervas How did relocating go? I know how life changes are a bit crazy and exciting. :)

**Comment by theodesp** [comment]
Given the extent of this ticket and the amount of code being written, it's more sensible to break it down into multiple tickets if it is to be merged at some point.

For example, this branch should be checked-out into multiple branches one for each subtask. When all the branches have been approved/tested then it should be merged to master.

**Comment by theodesp** [comment]
If its alright with you, I can rework this ticket as I don't like the inclusion of other repos code like that.
I suggest we split this task into 3 parts:

1. Implement Digest Auth Plugin parser and Setup code
2. Implement Nonce handling + Store
3. Implement Digest Auth handler + optional Apache style htdigest file handler

Thoughts?

**Comment by mholt** [comment]
I'm thinking more and more that this should be a plugin rather than a standard feature -- given its complexity, at least for now. Again, always easier to bring in rather than push out. In other words, an HTTP plugin with its own repo, that can be published to the Caddy website for optional inclusion in builds like the other ones.

But yeah, @theodesp, that process sounds good.

**Comment by hirusi** [comment]
Is there a community plugin for Digest Auth on Caddy v2? Or an official word?
